batcher/sweep_batch.go

```diff
@@ -0,0 +1,1330 @@
+func (b *Batch) publishBatchCoop(ctx context.Context) (btcutil.Amount,
+	error, bool) {
+
+	b.sweepsLock.Lock()
+	defer b.sweepsLock.Unlock()
+
+	batchAmt := btcutil.Amount(0)
+	sweeps := make([]Sweep, 0, len(b.sweeps))
+	fee := btcutil.Amount(0)
+	var weightEstimate input.TxWeightEstimator
+	addrOverride := false
+
+	// Create the batch transaction.
+	batchTx := &wire.MsgTx{
+		Version:  2,
+		LockTime: uint32(b.currentHeight),
+	}
+
+	for _, sweep := range b.sweeps {
+		// Append this sweep to an array of sweeps. This is needed to
+		// keep the order of sweeps stored, as iterating the sweeps map
+		// does not guarantee same order.
+		sweeps = append(sweeps, sweep)
+	}
+
+	// Add all the sweeps to the batch transaction.
+	for _, sweep := range sweeps {
+		if sweep.IsExternalAddr {
+			addrOverride = true
+		}
+
+		// Keep track of the total amount this batch is sweeping back.
+		batchAmt += sweep.Value
+
+		// Add this sweep's input to the transaction.
+		batchTx.AddTxIn(&wire.TxIn{
+			PreviousOutPoint: sweep.Outpoint,
+		})
+
+		err := sweep.HtlcSuccessEstimator(&weightEstimate)
+		if err != nil {
+			return fee, err, false
+		}
+	}
+
+	var destAddr btcutil.Address
+
+	if addrOverride {
+		// Sanity check, there should be exactly 1 sweep in this batch.
+		if len(sweeps) != 1 {
+			b.log.Warnf("non-wallet sweep batched with " +
+				"other sweeps")
+		}
+
+		destAddr = sweeps[0].DestAddr
+	} else {
+		var err error
+
+		// Generate a wallet address for the batch transaction's output.
+		destAddr, err = b.wallet.NextAddr(
+			ctx, "", walletrpc.AddressType_TAPROOT_PUBKEY, false,
+		)
+		if err != nil {
+			return fee, err, false
+		}
+	}
+
+	batchPkScript, err := txscript.PayToAddrScript(destAddr)
+	if err != nil {
+		return fee, err, false
+	}
+
+	weightEstimate.AddP2TROutput()
+
+	totalWeight := int64(weightEstimate.Weight())
+
+	fee = b.RbfCache.FeeRate.FeeForWeight(totalWeight)
+
+	// Add the batch transaction output, which excludes the fees paid to
+	// miners.
+	batchTx.AddTxOut(&wire.TxOut{
+		PkScript: batchPkScript,
+		Value:    int64(batchAmt - fee),
+	})
+
+	packet, err := psbt.NewFromUnsignedTx(batchTx)
+	if err != nil {
+		return fee, err, false
+	}
+
+	prevOuts := make(map[wire.OutPoint]*wire.TxOut)
+
+	for i, sweep := range sweeps {
+		txOut := &wire.TxOut{
+			Value:    int64(sweep.Value),
+			PkScript: sweep.Htlc.PkScript,
+		}
+
+		prevOuts[sweep.Outpoint] = txOut
+		packet.Inputs[i].WitnessUtxo = txOut
+	}
+
+	var psbtBuf bytes.Buffer
+	err = packet.Serialize(&psbtBuf)
+	if err != nil {
+		return fee, err, false
+	}
+
+	prevOutputFetcher := txscript.NewMultiPrevOutFetcher(prevOuts)
+
+	// Attempt to cooperatively sign the batch tx with the server.
+	err = b.coopSignBatchTx(ctx, packet, sweeps, prevOutputFetcher,
+		prevOuts, psbtBuf)
+	if err != nil {
+		return fee, err, false
+	}
+
+	b.log.Debugf("attempting to publish coop tx with feerate=%v, "+
+		"totalfee=%v, sweeps=%v, destAddr=%s", b.RbfCache.FeeRate, fee,
```

Comment:
I think it'd be cool to log a bit more data about the sweeps being published. Like maybe a `pkScript, value` pair for each or a swap hash. Could be useful down the line for debugging. (Same for the other function).

Reply:
will add this print only once at the end of the `publish` method instead